### PR TITLE
Added get functions and corresponding tests for vmc_vm_disk module

### PR DESCRIPTION
**Pull request by shaikres**
Added get functions and corresponding tests for vmc_vm_disk module

---

TODO:

- [x] - add integration tests
- [x] - get #215 merged

### CODE REVIEW

**Review by waynew — CHANGES_REQUESTED** [review]
I'm not sure if the SDDC is down or what, but when I try to run this by way of the CLI Example I just get 

```
requests.exceptions.ConnectTimeout: HTTPSConnectionPool(host='vcenter.sddc-10-191-252-30.vmwarevmc.com', port=443): Max retries exceeded with url: /api/session (Caused by ConnectTimeoutError(<urllib3.connection.HTTPSConnection object at 0x7f77b1fa8040>, 'Connection to vcenter.sddc-10-191-252-30.vmwarevmc.com timed out. (connect timeout=None)'))
```

**suggestion** at least one integration or functional test per module would  be a good idea. If anything causes this to fail correctly interacting with VMC, we want to know about it sooner rather than later.

File: src/saltext/vmware/modules/vmc_vm_disks.py [hunk 1]
```diff
@@ -0,0 +1,118 @@
...
+log = logging.getLogger(__name__)
+
+__virtualname__ = "vmc_vm_disks"
+
+
+def __virtual__():
+    return __virtualname__
+
+
+def get(hostname, username, password, vm_id, verify_ssl=True, cert=None):
```
Comment (waynew) [inline src/saltext/vmware/modules/vmc_vm_disks.py:22]: **suggestion** this gets the list of disks, correct? If so, this should be `list_`, with `__func_alias__ = {"list_": "list"}`

**Review by waynew — CHANGES_REQUESTED** [review]
In addition to the integration tests - how are we supposed to get the ID of the VMC VM?

I checked the docs at https://docs.saltproject.io/salt/extensions/salt-ext-modules-vmware/en/latest/vmc.html but was unable to find any information about how I would get that necessary information.

**Comment by shaikres** [comment]
to get the ID of the VMC VMs, we need to call the function: vmc_sddc.get_vms_by_sddc_id
this function is part of : https://github.com/saltstack/salt-ext-modules-vmware/pull/215

sample response : `[
        {
            "memory_size_MiB": 24576,
            "vm": "vm-36",
            "name": "NSX-Manager-0",
            "power_state": "POWERED_ON",
            "cpu_count": 4,
        },
        {
            "memory_size_MiB": 8192,
            "vm": "vm-37",
            "name": "NSX-Edge-0",
            "power_state": "POWERED_ON",
            "cpu_count": 4,
        },
    ]`
    
    we need to pick the "vm" field to pass as vm_id for getting the disks attached to vm

File: src/saltext/vmware/modules/vmc_vm_disks.py (hunk 1 above)
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_vm_disks.py:22]: This function returns the list of IDs of disks attached to the given vm
the backend API of this function returns the response as:
image

**Comment by shaikres** [comment]
Added other required functions, unit tests and integration tests.

Note: Steps to get  value for ``vm_id`` is mentioned as part of [PR-223](https://github.com/saltstack/salt-ext-modules-vmware/pull/223)

File: src/saltext/vmware/modules/vmc_vm_disks.py (hunk 1 above)
Comment (waynew) [reply src/saltext/vmware/modules/vmc_vm_disks.py:22]: This is still an issue

File: tests/integration/modules/test_vmc_vm_disks.py [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+@pytest.fixture
+def get_vm_disks(vm_disk_url, common_data, request_headers):
+    session = requests.Session()
+    response = session.get(
+        url=vm_disk_url,
+        verify=common_data["cert"] if common_data["verify_ssl"] else False,
+        headers=request_headers,
+    )
+    response.raise_for_status()
+    return response.json()
```
Comment (waynew) [inline tests/integration/modules/test_vmc_vm_disks.py:45]: **suggestion** I think that this test will suffer from the same problems that we fixed on the other test. For an integration test it doesn't really make sense to do out-of-band creation, especially since it looks like we're just doing CRUD ops here.

I was chatting about this kind of thing with Caleb on the Salt Core team and with the integration tests, when it's just doing CRUD ops makes sense to do something like this:


```
def test_vm_disks_smoke_test(...):  # maybe there are parameters or fixtures that it should use. Perhaps to generate a random disk name?

    existing_disk = salt_call_cli.run("vmc_vm_disks.get_by_id")
    assert existing_disk is None  # or whatever means it's not there
    
    existing_disks = salt_call_cli.run("vmc_vm_disks.list").json
    assert "some disk id" not in existing_disks  # or something similiar
    
    result = salt_call_cli.run("vmc_vm_disks.create", ...)
    assert result is True  # or whatever
    
    existing_disk = salt_call_cli.run("vmc_vm_disks.get_by_id", expected_id)
    assert existing_disk is not None
    assert existing_disk.id == expected_id
    
    result = salt_call_cli.run("vmc_vm_disks.update", ...)   # some values that you can update
    assert result is True  # or whatever the result should look like
    
    result = salt_call_cli.run("vmc_vm_disks.delete", ... )  # whatever the delete looks like
    
    existing_disk = ...  # Try to get the disk again, now that we've deleted it
    assert existing_disk is None  # or whatever it is that it looks like
```

Basically, since there's not really any logic here - it's just Create, Read, Update, Delete our pattern looks like this:

- Read a non-existent item. Assert that nothing was found
- Create the item brand new. Assert the creation result is correct
- Update the item. Assert the update returned the correct response
- Get the item again. Assert that the correct values have updated.
- Delete the item. Assert that the deletion returns the correct response.
- Get the item again. Assert that it's missing (or whatever the correct state is, for some of them that delete but it doesn't actually delete right away).    

Obviously, this should all change if there's actual logic in the module or state. But if it's basic CRUD? Yeah, that pattern is good :+1:

**Review by waynew — CHANGES_REQUESTED** [review]

File: src/saltext/vmware/modules/vmc_vm_disks.py (hunk 1 above)
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_vm_disks.py:22]: updated

File: tests/integration/modules/test_vmc_vm_disks.py (hunk 2 above)
Comment (shaikres) [reply tests/integration/modules/test_vmc_vm_disks.py:45]: updated tests according to the suggested pattern

**Comment by shaikres** [comment]
addressed the review comments and updated the code.

File: src/saltext/vmware/modules/vmc_vm_disks.py [hunk 3]
```diff
@@ -0,0 +1,480 @@
...
+
+            .. code::
+
+                "backing": {
+                    "type": "VMDK_FILE",
+                    "vmdk_file": "[WorkloadDatastore] 8dff8760-b6da-24d2-475b-0200a629f7fc/test-vm-1.vmdk"
+                }
+
+    new_vmdk
+        (optional) Specification for creating a new VMDK backing for the virtual disk.
```
[the remark this answers is not in the document]
Comment (waynew) [reply src/saltext/vmware/modules/vmc_vm_disks.py:329]: **question** Would a typical VMC user expect to use backing vs vmdk? This feels awkward to me. I would just expect to provide the criteria that are required. In fact, since there's no other type available for backing why not just have the API just look something like this?

```
create
   (Optional) whether or not to create the vmdk_file if it does not exist. Defaults to True.

vmdk_file
    (Optional) A base name for the physical resource backing for the disk, or an existing vmdk file. If the vmdk_file does not exist, it will be created, unless ``create`` is False. If ``vmdk_file`` is not provided, the name will be chosen by the server, derived from the name of the virtual machine.
     
capacity
    (Optional) The capacity for the disk, in bytes. Ignored if the vmdk file already exists.

storage_policy_id
    (Optional) The identifier for the storage policy to be used when creating the vmdk file.
```

Maybe there's a good reason to do it this way, but if we can simplify the options we should.

**Review by waynew — CHANGES_REQUESTED** [review]
I haven't tried out update & delete yet - but I've got a few questions/feedback items here.

**Comment by waynew** [comment]
Does look like the tests are passing now, though :+1:

File: src/saltext/vmware/modules/vmc_vm_disks.py [hunk 4]
```diff
@@ -0,0 +1,480 @@
...
+    verify_ssl
+        (optional) Option to enable/disable SSL verification. Enabled by default.
+        If set to False, the certificate validation is skipped.
+
+    cert
+        (optional) Path to the SSL client certificate file to connect to VMC Cloud Console.
+        The certificate can be retrieved from browser.
+
+    backing
+        (optional) Existing physical resource backing for the virtual disk.
```
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_vm_disks.py]: updated as per suggestion

File: src/saltext/vmware/modules/vmc_vm_disks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_vm_disks.py]: > **suggestion** same comment about optional. This is required unless `backing` is provided.

updated

[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_vm_disks.py]: > **question** Would a typical VMC user expect to use backing vs vmdk? This feels awkward to me. I would just expect to provide the criteria that are required. In fact, since there's no other type available for backing why not just have the API just look something like this?
> 
> ```
> create
>    (Optional) whether or not to create the vmdk_file if it does not exist. Defaults to True.
> 
> vmdk_file
>     (Optional) A base name for the physical resource backing for the disk, or an existing vmdk file. If the vmdk_file does not exist, it will be created, unless ``create`` is False. If ``vmdk_file`` is not provided, the name will be chosen by the server, derived from the name of the virtual machine.
>      
> capacity
>     (Optional) The capacity for the disk, in bytes. Ignored if the vmdk file already exists.
> 
> storage_policy_id
>     (Optional) The identifier for the storage policy to be used when creating the vmdk file.
> ```
> 
> Maybe there's a good reason to do it this way, but if we can simplify the options we should.

Yes the user is expected to use either backing or new_vmdk.  if he wants to use any existing vmdk file, he can pass value to the `backing` param ( that contains the type and path to that physical backing resource).
if he wants to use new physical resource, he can pass value to the `new_vmdk` (with the required specifications. we can also pass empty json, then vcenter will create the physical backing resource with default specifications ).

**Comment by shaikres** [comment]
addressed the suggestions.

File: src/saltext/vmware/modules/vmc_vm_disks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (waynew) [reply src/saltext/vmware/modules/vmc_vm_disks.py]: Since these are mutually exclusive, and unambiguous we should make this change - I'd be happy to pair on Wednesday morning my time/evening your time.

[the remark this answers is not in the document]
Comment (shaikres) [reply src/saltext/vmware/modules/vmc_vm_disks.py]: sure @waynew, i will schedule call for wednesday 7:30 pm IST

**Review by waynew — APPROVED** [review]
Tests are all passing :rocket:
